spec/models/vocabulary_spec.rb

```diff
@@ -39,6 +43,130 @@
+        )
+      end
+
+      it 'returns validation error' do
+        expect(vocabulary.save).to be false
+        expect(vocabulary.errors.full_messages).to include 'Custom fields field_key cannot be a reserved field name'
+      end
+    end
+
+    context 'when custom_field field_key starts contains invalid characters' do
+      let(:vocabulary) do
+        FactoryBot.build(
+          :vocabulary, custom_fields: { AUTHORITY: { data_type: 'string', label: 'Authority' } }
+        )
+      end
+
+      it 'returns validation error' do
+        expect(vocabulary.save).to be false
+        expect(vocabulary.errors.full_messages).to include 'Custom fields field_key can only contain lowercase alphanumeric characters and underscores'
+      end
+    end
+
+    context 'when custom_field missing label' do
+      let(:vocabulary) do
+        FactoryBot.build(
+          :vocabulary, custom_fields: { classification: { data_type: 'string' } }
+        )
+      end
+
+      it 'returns validation error' do
+        expect(vocabulary.save).to be false
+        expect(vocabulary.errors.full_messages).to include 'Custom fields each custom_field must have a label and data_type defined'
+      end
+    end
+
+    context 'when custom_field missing data_type' do
+      let(:vocabulary) do
+        FactoryBot.build(
+          :vocabulary, custom_fields: { classification: { label: 'Classification' } }
+        )
+      end
+
+      it 'returns validation error' do
+        expect(vocabulary.save).to be false
+        expect(vocabulary.errors.full_messages).to include(
+          'Custom fields each custom_field must have a label and data_type defined',
+          'Custom fields data_type must be one of string, number or boolean'
+        )
+      end
+    end
+
+    context 'when a custom_field\'s data_type is invalid' do
+      let(:vocabulary) do
+        FactoryBot.build(
+          :vocabulary, custom_fields: { classification: { label: 'Classification', data_type: 'decimal' } }
+        )
+      end
+
+      it 'returns validation error' do
+        expect(vocabulary.save).to be false
+        expect(vocabulary.errors.full_messages).to include 'Custom fields data_type must be one of string, number or boolean'
+      end
+    end
+  end
+
+  describe '#add_custom_field' do
+    let(:vocabulary) { FactoryBot.create(:vocabulary) }
+
+    context 'when field_key is blank' do
+      it 'raises an error' do
+        expect { vocabulary.add_custom_field() }.to raise_error 'field_key cannot be blank'
+      end
+    end
+
+    context 'when field_key already exists' do
+      before do
+        vocabulary.add_custom_field(field_key: 'classification', label: 'classification', data_type: 'string')
+        vocabulary.save
+      end
+
+      it 'raises an error' do
+        expect { vocabulary.add_custom_field(field_key: 'classification') }.to raise_error 'field_key cannot be added because its already a custom field'
```

Comment:
```suggestion
        expect { vocabulary.add_custom_field(field_key: 'classification') }.to raise_error 'field_key cannot be added because it\'s already a custom field'
```